### PR TITLE
Allow custom currency icons

### PR DESCRIPTION
**Pull request by Larkinabout**
Allow for users/modules to customise the currency icons by adding the icon to the config and replacing the hardcoded CSS with a style tag for the background url. Related to #3145.

### CODE REVIEW

**Comment by Fyorl** [comment]
This is a bit more complicated, unfortunately, because there are journals out there that use these currency icon styles, as well as the award enricher I think.

There might be less complicated/ugly options, but we could make these styles entirely dynamic.

```js
Hooks.on("setup", () => {
  const style = document.createElement("style");
  style.innerHTML = `
    :is(.dnd5e2, .dnd5e2-journal) :is(i, span).currency {
      ${Object.entries(CONFIG.DND5E.currencies).reduce((rules, [k, { icon }]) => {
        return `${rules}\n.${k} { background-image: url("${icon}"); }`;
      }, "")}
    }
  `;
  document.head.append(style);
});
```

**Comment by Larkinabout** [comment]
Sounds good to me.

File: less/v2/apps.less [hunk 1]
```diff
@@ -664,12 +664,6 @@
   height: 1em;
   display: inline-block;
   background: transparent no-repeat center / contain;
-
-  &.cp { background-image: url("icons/currency/copper.webp"); }
-  &.sp { background-image: url("icons/currency/silver.webp"); }
-  &.ep { background-image: url("icons/currency/electrum.webp"); }
-  &.gp { background-image: url("icons/currency/gold.webp"); }
-  &.pp { background-image: url("icons/currency/platinum.webp"); }
```
Comment (arbron) [inline less/v2/apps.less:672]: We use these in some journals, such as in Phandelver & Below, for explaining currencies, so they shouldn't be removed.

File: module/applications/award.mjs [hunk 2]
```diff
@@ -57,8 +57,13 @@ export default class Award extends DialogMixin(FormApplication) {
...
     context.CONFIG = CONFIG.DND5E;
-    context.currency = Object.entries(CONFIG.DND5E.currencies).reduce((obj, [k, { label }]) => {
-      obj[k] = { label, value: this.options.currency ? this.options.currency[k] : this.object?.system.currency[k] };
+    context.currency = Object.entries(CONFIG.DND5E.currencies).reduce((obj, [k, { label, icon }]) => {
+      obj[k] = {
+        label,
+        icon,
+        value: this.options.currency
+          ? this.options.currency[k]
+          : this.object?.system.currency[k] };
```
Comment (arbron) [inline module/applications/award.mjs:66]:
```suggestion
    context.currency = Object.entries(CONFIG.DND5E.currencies).reduce((obj, [k, { label, icon }]) => {
      obj[k] = {
        label, icon,
        value: this.options.currency ? this.options.currency[k] : this.object?.system.currency[k]
      };
```

File: module/config.mjs [hunk 3]
```diff
@@ -1404,27 +1404,32 @@ DND5E.currencies = {
   pp: {
```
Comment (arbron) [inline module/config.mjs:1404]: Be sure to update the type definition above to document the added icon:

```
 * @property {string} icon          Icon representing the currency in the interface.
```

File: templates/apps/currency-manager.hbs [hunk 4]
```diff
@@ -21,7 +21,7 @@
                 {{#each currency}}
                 <label>
                     {{#with (lookup (lookup @root.CONFIG.currencies @key) "label") as |label|}}
-                    <i class="currency {{@key}}" data-tooltip="{{label}}" aria-label="{{label}}"></i>
+                    <i class="currency" style="background-image: url('{{ lookup (lookup @root.CONFIG.currencies @key) "icon" }}');" data-tooltip="{{label}}" aria-label="{{label}}"></i>
                     {{/with}}
```
Comment (arbron) [inline templates/apps/currency-manager.hbs:25]:
```javascript

                    {{#with (lookup @root.CONFIG.currencies @key) as |currency|}}
                    <i class="currency" style="background-image: url('{{ lookup currency "icon" }}');"
                       data-tooltip="{{ lookup currency 'label' }}" aria-label="{{ lookup currency 'label' }}"></i>
                    {{/with}}
```

**Review by arbron — CHANGES_REQUESTED** [review]
You'll also need to fix the award enricher to use the correct icons.

**Comment by Larkinabout** [comment]
So, forget making them dyanmic, leave the styles as is and replace the references within the dnd5e system to use the config icon?

It should mean that if currencies are removed from the config, everything that points to those styles will still work. With the understanding that a journal won't reflect homebrew currencies, which I don't think anyone would expect anyway.

File: less/v2/apps.less (hunk 1 above)
Comment (Larkinabout) [reply less/v2/apps.less:769]: Added back in.

**Review by arbron — APPROVED** [review]

**Pull request merged**